k2/csrc/fsa_algo.h

```diff
@@ -158,6 +163,21 @@ Fsa LinearFsa(const Array1<int32_t> &symbols);
  */
 FsaVec LinearFsas(Ragged<int32_t> &symbols);
 
+/* Compute the shortest distance in the tropical semiring, but
+   with the opposite sign. That is, it uses `max` instead of `min.
+
+   @param [in] fsa           The input FSA. It is on CPU and must be
+                             topologically sorted.
+   @param [out] arc_map      The arc indexes of the best path from the start
+                             state to the final state. It is empty if there
+                             is no such path.
+   @return The shortest distance from the start state to the final state.
+           Returns negative infinity if there is no such path.
+ */
+double ShortestDistance(Fsa &fsa, Array1<int32_t> *arc_map = nullptr);
+
+void ShortestPath(Fsa &src, Fsa *out, Array1<int32_t> *arc_map = nullptr);
```

Comment:
BTW, I think we may not need this version `ShortestPath`, `ShortestDistance` with `arc_map` is good enough. @danpovey what do you think?

Reply:
I think this function is useful enough to keep, since there isn't, I believe, a simple way to create an FSA from an FSA and an arc_map.
Incidentally, in ShortestDistance, I don't think we should call the list of arcs `arc_map` as it doen't really map from anything specific.  I think maybe `best_path_arcs` would make more sense.  In fact, the 2 functions could be merged into ShortestPath and the fsa output made optional.

Eventually we can add an efficient implementation for FsaVec on GPU.  Its core will be similar to GetForwardScores, except it will keep track of the best traceback.

Reply:
> I think this function is useful enough to keep, since there isn't, I believe, a simple way to create an FSA from an FSA and an arc_map.

The implementation, I believe, is just to create an Fsa from a list of arcs (instead of create an Fsa from an FSA and an arc_amp), that's why I said we may not need this version.
https://github.com/k2-fsa/k2/blob/9147dbcee8295bcce4e03e400a92dd1d3334aced/k2/csrc/fsa_algo.cu#L338

Thus I think we should define the interface either as
1. return arc_map in the best path, i.e. Array1<int32_t> for arc_map, or as
2. return an linear Fsa which is the best path.

I think we can just create a function to create an FSA from an src FSA and an arc_map, so the interface here could be just returning arc_map.

Reply:
Sure.  Bear in mind that eventually we'll need this to work for FsaVec, and in that case ShortestDistance would have to output a `Ragged<int32_t>`; and we'd need a function (working on GPU!) to create an FsaVec from an input FsaVec and a `Ragged<int32_t>`.  That's very easy to do.

Reply:
Dan, why do we need to return `Ragged<int32_t>` here as arc_map, instead of `Array1<int32_t>`? For FsaVec version, what index is in the returned arc_map? idx2 or idx012?

Reply:
I mean `best_arc_paths`, as it's not arc_map.

Reply:
>  what index is in the returned arc_map? idx2 or idx012?

I think the output argument `Ragged<int32_t> *best_path_arcs ` will have two axes containing:

- fsa 0, arc indexes [x x x x]
- fsa 1, arc indexes [x x x]
- ...

Reply:
mm. what `x` here it is? if it's arc-index in the original FsaVec,why not just return an Array1 with `x` is `arc_idx012`

Reply:
It's for a FsaVec.

Reply:
Mm.  I still think the right interface is to have it output a Ragged<T>.   (E.g. it makes it easier to look up the output for an individual FSA).  We'd have to have a separate version for FsaVec vs. Fsa, if we support Fsa.  It makes me think maybe we should have supported a Ragged<T> with only one axis, but that's not urgent.

Reply:
... the Fsa version, i.e. the GPU implementation, could just add one extra axis and then forward the contents of the FsaVec version of the output (the elements of the Ragged<int32_t>).

Reply:
Do you mean the interface should look like the following:

```cpp
Ragged<int32_t> ShortestPath(FsaVec &src);

FsaVec CreateFsaVec(FsaVec &src, Ragged<int32_t> &arc_indexes);
```

`arc_indexes` has two axes:
- fsa 0, [arc_idx012 of fsa 0]
- fsa 1, [arc_idx012 of fsa 1]
- ...

`Ragged<int32_t>` is preferred over `Array1<int32_t>` since it is much easier to get the number of states per FSA
and to distribute the work of rebuilding the final FsaVec to different threads.

Reply:
Yes that's what I mean.   However I think the function should optionally take the 'entering_arcs' vector that's output by GetForwardScores(), as this would be more friendly to future use of caching code.  If it is not supplied, the function can create it internally (and probably then recurse to itself).